crates/transaction-tracing/src/tracing.rs

```diff
@@ -0,0 +1,150 @@
+use alloy_primitives::TxHash;
+use eyre::Result;
+use futures::StreamExt;
+use lru::LruCache;
+use reth::api::{BlockBody, FullNodeComponents};
+use reth::core::primitives::{AlloyBlockHeader, SignedTransaction};
+use reth::transaction_pool::{FullTransactionEvent, TransactionPool};
+use reth_exex::{ExExContext, ExExEvent, ExExNotification};
+use reth_tracing::tracing::debug;
+use std::num::NonZeroUsize;
+use std::time::Instant;
+use tokio_stream::wrappers::ReceiverStream;
+
+/// Types of transaction events to track
+#[derive(Debug)]
+enum TxEvent {
+    Dropped,
+    Replaced,
+    BlockInclusion,
+}
+
+/// Simple ExEx that tracks transaction timing from mempool to inclusion
+struct Tracker {
+    /// Map of transaction hash to timestamp when first seen in mempool
+    txs: LruCache<TxHash, Instant>,
+}
+
+impl Tracker {
+    fn new() -> Self {
+        Self {
+            txs: LruCache::new(NonZeroUsize::new(20000).unwrap()),
+        }
+    }
+
+    fn transaction_inserted(&mut self, tx_hash: TxHash) {
+        let now = Instant::now();
+        self.txs.put(tx_hash, now);
+        debug!(target: "transaction-tracing", tx_hash = ?tx_hash, "Transaction added to mempool");
+    }
+
+    fn transaction_included_in_block(&mut self, tx_hash: TxHash, block_number: u64) {
+        if self.txs.contains(&tx_hash) {
+            self.transaction_event(tx_hash, TxEvent::BlockInclusion);
+        } else {
+            debug!(
+                target: "transaction-tracing",
+                tx_hash = ?tx_hash,
+                block_number = ?block_number,
+                "Transaction included in block (not tracked by ExEx)",
+            );
+        }
+    }
```

Comment:
nit: personally I would combine this with transaction event, you could move the else clause there (drop the block number param) and delete this function